### DISCUSSION TITLE
Improve behaviour of seek_in for text mode files

### DISCUSSION THREAD

**Discussion by dra27** [ocaml/ocaml#11958] [Features]
The incorrect behaviour reported in #9868 and fixed in #9872 assumed that no program could have relied on it (https://github.com/ocaml/ocaml/pull/9872#issuecomment-688679715). It turns out one of the tests inadvertently relied on that behaviour (fixed in #10108), but I wonder if this is worth a little further thought and possibly a further tweak. In particular, the result of the new behaviour on this previously working program was to produce garbage and, even having reviewed the change, I needed to bisect the compiler to identify the cause (especially as in this case the test program doesn't directly use `pos_in`).

Prior to #9872 `pos_in` and `seek_in` actually work correctly on text files as long as they never overflow the channel's buffer. That buffer is 64KiB which is a quite large number of "text" files! I'm wondering if any of these additional behaviours could be considered instead of the present behaviour which doesn't have any kind of warnings on _existing_ code:

- Prohibit `seek_in` completely on text files by raising an exception. More breaking change, but it's much clearer what's gone wrong. Very simple.
- Allow `seek_in` (and possibly `seek_out`) only if the buffer has never been flushed. Still simple. Bit weird.
- Use a larger channel buffer to text files with the previous solution (or allow the buffer to grow for text files). Slightly less weird; increases complexity.

As an aside, I think the correct handling for text mode in the stdlib would be:
- Deprecate `open_in`, because text mode for input is a totally flawed idea (a CRLF file should read in the same way on Unix and Windows)
- Make `input_line` read to the next `\r*\n`/EOF
- Have `seek_out` and `pos_out` raise `Invalid_argument` for text mode files

**Comment by xavierleroy** [comment]
With the current behavior on text files, it is possible to reliably `seek_in` or `seek_out` to 0 or to positions returned by `in_channel_length` and `out_channel_length`.  That's slightly more useful than prohibiting `seek_in` on text files entirely.

I did not realize that the old behavior sort of works on files small enough to fit in the buffer.  It feels  like an accident, not something to rely on.  But this can be argued.

Deprecating text mode for input but not for output sounds weird.

Another solution would be to do the CRLF conversion in the OCaml runtime, not relying on the CRT to do it.  The buffers would contain "binary" data, with CRLF end of lines.  But that's a lot of code to add to the runtime system.

Even in this case, code that uses lexer positions as arguments to `seek_in` and `seek_out` over a text file would still not work.

**Comment by dra27** [comment]
> Deprecating text mode for input but not for output sounds weird.

Indeed, but the point - assuming we don't implement text mode ourselves - is that OS-implemented text mode for input is broken, because it results in non-portable programs (amusingly, only Windows works: it's the only platform able to reliably read both LF and CRLF files as intended). OS-implemented text mode for output is not broken, since it implies "write a text file that's the default for this platform"

**Comment by dra27** [comment]
> Another solution would be to do the CRLF conversion in the OCaml runtime, not relying on the CRT to do it. The buffers would contain "binary" data, with CRLF end of lines. But that's a lot of code to add to the runtime system.

Agreed, it's a lot of code - I'd prefer we moved towards having higher level functions which allowed manipulating text files more naturally, rather than "magic" in the channel. For example, `input_line` always ignoring `\r` and potentially channels automatically recording whether the data has been CRLF or LF-formatted.

**Comment by dra27** [comment]
This issue is part of a bigger picture for reducing surprising differences between Unix and Windows; we're all on other things at the moment, but I'd like to keep it open.

**Comment by dra27** [comment]
I would still like to oversee/make some improvements to the standard library for this. WSL especially makes it very easy to have text files written as "OS-native" (for the host - i.e. Windows) read "OS-native" by WSL (i.e. Linux) all within the same file system. I think since the addition of 4.14's `In_channel` and `Out_channel` modules, this might be something which can be more easily done (possibly even like the `LargeFile` module variants, etc.) and certainly could be done while leaving strictly compatible versions of the functions in `Stdlib` in place.
